Migration script for old articles, fix #195

davidbgk opened the pull request:
Note: it drops fixtures for articles/users so you will not be able anymore to login with @example.com accounts.

davidbgk committed:
Migration script for old articles, fix #195

Note: it drops fixtures for articles/users so you will not be able anymore to login with @example.com accounts.

davidbgk committed:
Merge branch 'master' into 195-import-article-base

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -25,11 +25,15 @@ class Article(db.Document, UploadableImageMixin):
     tags = db.ListField(db.ReferenceField(Tag, reverse_delete_rule=PULL))
+    old_pk = db.IntField()  # In use for migrations (references in related).
+    old_category_slug = db.StringField()  # In use for redirections.
+    old_slug = db.StringField()  # In use for redirections.
```
create an `archive` property with these 3 fields inside?
Could even be an _EmbeddedDocument_ :)

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -25,11 +25,15 @@ class Article(db.Document, UploadableImageMixin):
     meta = {
-        'allow_inheritance': True
+        'allow_inheritance': True,
+        'indexes': ['old_pk']
```
In which case we'd have `'indexes': ['achive.pk']`

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -1,43 +1,174 @@
 
+PROGRESSIST_TEMPLATE = ('{prefix} {animation} {percent} ({done}/{total}) '
+                        'ETA: {eta:%H:%M} {elapsed}')
+NOW = datetime.now()
```
Please remind me why not storing `utcnow()` instead?

davidbgk replied:
That date is not stored, just compared to exclude 3-months old content. I guess we are not about an hour precision 😉

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -1,43 +1,174 @@
+        normalized_language = 'de'
+    elif normalized_language == 'sp':
+        normalized_language = 'es'
+    return normalized_language
```
Bewarn that _pol_ is in the list of possible languages in the old system config.
Not sure if it's used though.

vinyll reviewed (commented):
💣

davidbgk committed:
Turn archives data into an embedded document

Refs #195

davidbgk committed:
Handle galleries and old urls for redirections

Refs #195

davidbgk committed:
Handle non-ASCII slugs generation, refs #195

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -113,32 +117,51 @@ def handle_groups(groups):
+        image_fields = image['fields']
+        image_src = normalize_image(image_fields['image'])
+        body += f'<p><img src="{image_src}" alt="{image_fields["title"]}"></p>'
+        body += image_fields['body']
```
Wondering if we shouldn't got for
```html
<figure>
  <img src="{image_src}" alt="{image_fields["title"]}">
  <figcaption>{image_fields['body']}</figcaption>
</figure>
```

It could be easier to css tune old galleries later.

davidbgk replied:
Way better, thanks mate 👍

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -113,32 +117,51 @@ def handle_groups(groups):
                 'summary': group_fields['about'],
             }
+            if data['language'] == 'xx':
+                # It happens with unused tags(?)
```
🤢

vinyll commented on cafebabel/tests/test_articles.py:
```diff
@@ -25,6 +25,12 @@ def test_access_article_with_large_slug_should_return_200(client,
     assert response.status_code == HTTPStatus.OK
 
 
+def test_article_with_non_ascii_title(published_article):
```
😎

vinyll committed:
redirecting archived articles from Django to new version

davidbgk committed:
Add support for relateds, fix #197

davidbgk committed:
Move relateds to the embedded archive, refs #197

davidbgk committed:
Fallback on sane default for empty summary, refs #195

davidbgk commented on cafebabel/tests/test_articles_archives.py:
```diff
@@ -0,0 +1,32 @@
+    published_article.archive = ArticleArchive(id=1, url=url)
+    published_article.save()
+    response = client.get(url)
+    assert response.status_code == 301
```
Use HTTPStatus?

vinyll committed:
changed status code to httpstatus constants

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -135,8 +141,10 @@ def aggregate_gallery_body(gallery):
+        body += ('<figure>'
+                 f'<img src="{image_src}" alt="{image_fields["title"]}">'
+                 f'<figcaption>{image_fields["body"]}</figcaption>'
+                 '</figure>')
```
Other suggestion
```python
body += f'''<figure>
               <img…
         '''
```
More a matter of preference I suppose; I leave this to your consideration

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -178,8 +178,8 @@ def create_article(old_article):
         'title': fields['title'],
-        'summary': article_fields['original_header'] or 'TODO',
-        'body': article_fields['body'] or 'TODO',
+        'summary': article_fields['original_header'] or '',
```
When a summary was too long, it was ⬇️ moved as the  1️⃣st paragraph into the body and **bolded**.

Shall we ditch this case or treat it?

davidbgk replied:
It's not always the case :/

vinyll commented on cafebabel/articles/models.py:
```diff
@@ -12,6 +12,12 @@
+class ArticleArchive(db.EmbeddedDocument):
+    pk = db.IntField()  # In use for migrations (references in related).
+    url = db.StringField()  # In use for redirections.
+    relateds = db.ListField()
```
I see you had put this in the `Article` object in the first place before moving it here.
I shared the opinion of the `Article` property as we will need it the same way for all articles.
What made you change for this?

davidbgk replied:
See my other comment, here we just store old ids for later use.

vinyll commented on cafebabel/django_migrations.py:
```diff
@@ -1,43 +1,224 @@
+        'archive': ArticleArchive(
+            pk=old_article['pk'],
+            url=old_article['url'],
+            relateds=fields['relateds']
```
Shouldn't we save the new reference instead?

I was thinking that performance-wise it would be consuming but I think you make 2 rounds anyways.

I suppose if you decide to keep this property in the Archive it makes as much sense to keep the archive _pk_ 🤔

davidbgk replied:
I think we have to make a third round to be really complete and given that we don't need those yet I prefer to store row old ids and to add references to the `Article` when we will implement the feature for all articles.

davidbgk committed:
Remove HTML from old titles, refs #195

davidbgk committed:
Allow HTML summaries in templates, refs #195

davidbgk committed:
Only allow a restricted set of languages, refs #195

davidbgk committed:
Merge branch 'master' into 195-import-article-base

davidbgk commented on cafebabel/core/views.py:
```diff
@@ -11,3 +17,11 @@ def home():
     return send_from_directory(current_app.config['UPLOADS_FOLDER'], filename)
+
+
+@cores.route('/<regex("[a-z-]+"):_tag>/<regex("[a-z]+"):_>/<_slug>.html')
```
@vinyll What about turning `<regex("[a-z]+"):_>` into `article`? Are there other cases?

davidbgk committed:
Merge branch 'master' into 195-import-article-base

# Conflicts:
#	cafebabel/templates/articles/detail.html

davidbgk committed:
Split archives to their dedicated blueprint, refs #195

This way, we can fallback to archives routes if no other match is found for regular routes.

vinyll commented on cafebabel/users/models.py:
```diff
@@ -21,6 +20,10 @@ class UserProfile(db.EmbeddedDocument, UploadableImageMixin):
     old_pk = db.IntField()  # In use for migrations (references in articles).
 
+    meta = {
+        'indexes': ['old_pk']
```
Should the meta be on the Embedded class or the Article class?
https://stackoverflow.com/a/30455984/328117

davidbgk committed:
Appropriated index on user’s profile old pks, refs #195

vinyll commented:
🎈
